File: public/api/tags.php
<?php

require dirname(__DIR__, 2) . '/vendor/autoload.php';

use App\Database;
use App\TagRepository;

header('Content-Type: application/json');

$method = $_SERVER['REQUEST_METHOD'];
$db     = Database::get();
$repo   = new TagRepository($db);

if ($method === 'GET') {
    echo json_encode(['tags' => $repo->all()]);
    exit;
}

if ($method === 'POST') {
    $body = json_decode(file_get_contents('php://input'), true);
    $name = trim($body['name'] ?? '');

    if ($name === '') {
        http_response_code(400);
        echo json_encode(['error' => 'Tag name is required']);
        exit;
    }

    $existing = $repo->findByName($name);
    if ($existing) {
        echo json_encode(['id' => (int) $existing['id'], 'name' => $existing['name']]);
        exit;
    }

    $id = $repo->create($name);
    echo json_encode(['id' => $id, 'name' => $name]);
    exit;
}

if ($method === 'DELETE') {
    $id = (int) ($_GET['id'] ?? 0);
    if ($id === 0) {
        http_response_code(400);
        echo json_encode(['error' => 'Missing id']);
        exit;
    }
    $repo->delete($id);
    echo json_encode(['success' => true]);
    exit;
}

http_response_code(405);
echo json_encode(['error' => 'Method not allowed']);

File: internal/handler/tags.go
package handler

import (
	"imagemanager/internal/model"
	"imagemanager/internal/repo"
	"net/http"
	"strconv"
)

func (a *App) Tags(w http.ResponseWriter, r *http.Request) {
	tags := repo.NewTagRepo(a.DB)

	switch r.Method {
	case http.MethodGet:
		all, err := tags.All()
		if err != nil {
			jsonErr(w, 500, "Failed to load tags")
			return
		}
		if all == nil {
			jsonOK(w, map[string]any{"tags": []model.Tag{}})
			return
		}
		jsonOK(w, map[string]any{"tags": all})

	case http.MethodPost:
		var body struct {
			Name string `json:"name"`
		}
		if err := readJSON(r, &body); err != nil || body.Name == "" {
			jsonErr(w, 400, "Tag name is required")
			return
		}
		existing, err := tags.FindByName(body.Name)
		if err != nil {
			jsonErr(w, 500, "DB error")
			return
		}
		if existing != nil {
			jsonOK(w, map[string]any{"id": existing.ID, "name": existing.Name})
			return
		}
		id, err := tags.Create(body.Name)
		if err != nil {
			jsonErr(w, 500, "Failed to create tag")
			return
		}
		jsonOK(w, map[string]any{"id": id, "name": body.Name})

	case http.MethodDelete:
		idStr := r.URL.Query().Get("id")
		id, err := strconv.ParseInt(idStr, 10, 64)
		if err != nil || id == 0 {
			jsonErr(w, 400, "Missing id")
			return
		}
		if err := tags.Delete(id); err != nil {
			jsonErr(w, 500, "Failed to delete tag")
			return
		}
		jsonOK(w, map[string]bool{"success": true})

	default:
		jsonErr(w, 405, "Method not allowed")
	}
}
